metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java

```diff
@@ -1204,6 +1317,298 @@ void generateLeaderAndIsrUpdates(String context,
+        }
+    }
+
+    void cancelPartitionReassignment(String topicName,
+                                     TopicIdPartition topicIdPartition,
+                                     PartitionControlInfo partition,
+                                     List<ApiMessageAndVersion> records) {
+        if (!partition.isReassigning()) {
+            throw new NoReassignmentInProgressException(NO_REASSIGNMENT_IN_PROGRESS.message());
+        }
+        PartitionChangeRecord record = new PartitionChangeRecord().
+            setTopicId(topicIdPartition.topicId()).
+            setPartitionId(topicIdPartition.partitionId());
+        if (partition.removingReplicas.length > 0) {
+            record.setRemovingReplicas(Collections.emptyList());
+        }
+        if (partition.addingReplicas.length > 0) {
+            record.setAddingReplicas(Collections.emptyList());
+        }
+        RemovingAndAddingReplicas removingAndAddingReplicas =
+            new RemovingAndAddingReplicas(partition.removingReplicas, partition.addingReplicas);
+        List<Integer> currentReplicas = Replicas.toList(partition.replicas);
+        List<Integer> currentIsr = Replicas.toList(partition.isr);
+        List<Integer> revertedReplicas = removingAndAddingReplicas.
+            calculateRevertedReplicas(currentReplicas, currentIsr);
+        if (!revertedReplicas.equals(currentReplicas)) {
+            record.setReplicas(revertedReplicas);
+            List<Integer> newIsr = new ArrayList<>();
+            for (int replica : partition.isr) {
+                if (revertedReplicas.contains(replica)) {
+                    newIsr.add(replica);
+                }
+            }
+            if (!newIsr.equals(currentIsr)) {
+                if (!newIsr.contains(partition.leader)) {
+                    int newLeader = bestLeader(Replicas.toArray(revertedReplicas),
+                        Replicas.toArray(newIsr),
+                        configurationControl.uncleanLeaderElectionEnabledForTopic(topicName),
+                        r -> clusterControl.unfenced(r));
+                    if (!electionWasClean(newLeader, Replicas.toArray(newIsr))) {
+                        newIsr = Collections.singletonList(newLeader);
+                    }
+                    record.setLeader(newLeader);
+                }
+                record.setIsr(newIsr);
+            }
+        }
+        records.add(new ApiMessageAndVersion(record, (short) 0));
+    }
+
+    void changePartitionReassignment(String topicName,
+                                     TopicIdPartition part,
+                                     PartitionControlInfo partitionInfo,
+                                     ReassignablePartition partition,
+                                     List<ApiMessageAndVersion> records) {
+        // Check that the requested partition assignment is valid.
+        validateManualPartitionAssignment(partition.replicas(), OptionalInt.empty());
```

Comment:
I think we allow changing this now so that different partitions have different replica counts, so not allowing this for KRaft mode would be an incompatibility.

Reply:
Hmm, when did we allow different partitions in the same topic to have different replica counts? This will be weird for min.isr since it is set at the topic level.

Reply:
I think reassignment has always supported giving different partitions in the same topic different replica counts. Example:
```
$ ./bin/kafka-topics.sh --bootstrap-server localhost:9092 --describe
Topic: foo      TopicId: o3KToxKaTX2rB6d-VufJig PartitionCount: 2       ReplicationFactor: 2    Configs: segment.bytes=1073741824
        Topic: foo      Partition: 0    Leader: 0       Replicas: 0,1   Isr: 0,1
        Topic: foo      Partition: 1    Leader: 1       Replicas: 1,0   Isr: 1,0

$ cat /tmp/reassignment.json
{"partitions": [{"topic": "foo", "partition": 1, "replicas": [0]}], "version":1}

$ ./bin/kafka-reassign-partitions.sh --bootstrap-server localhost:9092 --reassignment-json-file /tmp/reassignment.json --execute

$ ./bin/kafka-topics.sh --bootstrap-server localhost:9092 --describe
Topic: foo      TopicId: o3KToxKaTX2rB6d-VufJig PartitionCount: 2       ReplicationFactor: 2    Configs: segment.bytes=1073741824
        Topic: foo      Partition: 0    Leader: 0       Replicas: 0,1   Isr: 0,1
        Topic: foo      Partition: 1    Leader: 0       Replicas: 0     Isr: 0
```
Should we file a JIRA to possibly fix this in a later version of Kafka? It does seem like it could create problems. But changing it would also break compatibility, of course.

Reply:
Thanks, Colin. It seems that we don't enforce the same replication factor within a topic now. So, we can keep the logic as it is and have a followup jira to improve this.